File: typescript/src/utils/errorHandler.ts
export interface ErrorDetails {
  category?: unknown;
  code?: unknown;
}

export interface ErrorResponse {
  code?: unknown;
  message?: unknown;
  detail?: unknown;
  request_id?: unknown;
  trace_id?: unknown;
  error?: ErrorDetails;
}

/** 统一解析模板错误体，并在迁移期兼容 legacy detail。 */
import { AxiosError, AxiosHeaders, isAxiosError } from 'axios';

export class NewNanManagerHttpError extends Error {
  public readonly statusCode?: number;
  public readonly code?: string;
  public readonly apiCode?: number;
  public readonly errorCategory?: string;
  public readonly machineCode?: string;
  public readonly requestId?: string;
  public readonly traceId?: string;
  public readonly retryAfter?: string;

  constructor(error: AxiosError<ErrorResponse>) {
    const data = error.response?.data;
    const message = typeof data?.message === 'string' && data.message.length > 0
      ? data.message
      : typeof data?.detail === 'string' && data.detail.length > 0
        ? data.detail
        : error.message;
    super(message);
    this.name = 'NewNanManagerHttpError';
    this.statusCode = error.response?.status;
    this.code = error.code;
    this.apiCode = typeof data?.code === 'number' && Number.isInteger(data.code) ? data.code : undefined;
    this.errorCategory = typeof data?.error?.category === 'string' ? data.error.category : undefined;
    this.machineCode = typeof data?.error?.code === 'string' ? data.error.code : undefined;
    const headers = error.response?.headers;
    const requestId = headers instanceof AxiosHeaders ? headers.get('X-Request-ID') : headers?.['x-request-id'];
    const retryAfter = headers instanceof AxiosHeaders ? headers.get('Retry-After') : headers?.['retry-after'];
    this.requestId = typeof requestId === 'string' ? requestId : typeof data?.request_id === 'string' ? data.request_id : undefined;
    this.traceId = typeof data?.trace_id === 'string' ? data.trace_id : undefined;
    this.retryAfter = typeof retryAfter === 'string' ? retryAfter : undefined;
  }
}

/**
 * 通用错误处理函数
 * 用作所有API的第三个参数（ErrorHandler）
 * 与现有的player模块中的错误处理逻辑保持一致
 */
export const commonErrorHandler = ({ error }: { error: unknown }): never => {
  if (isAxiosError<ErrorResponse>(error)) {
    throw new NewNanManagerHttpError(error);
  }
  throw error instanceof Error ? error : new Error(String(error));
};

File: golang/utils/error_handler.go
package utils

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/go-resty/resty/v2"
)

// ErrorDetails is the stable machine-readable error registry entry.
type ErrorDetails struct {
	Category string `json:"category"`
	Code     string `json:"code"`
}

// ErrorResponse is the template error response. Detail remains a migration
// fallback for servers that still emit the legacy body.
type ErrorResponse struct {
	Code      int          `json:"code"`
	Message   string       `json:"message"`
	Detail    string       `json:"detail"`
	RequestID string       `json:"request_id"`
	TraceID   string       `json:"trace_id"`
	Error     ErrorDetails `json:"error"`
}

// HTTPError preserves response metadata for errors.As and caller retry decisions.
type HTTPError struct {
	StatusCode  int
	Detail      string
	Code        int
	Category    string
	MachineCode string
	RequestID   string
	TraceID     string
	RetryAfter  string
}

func (e *HTTPError) Error() string {
	return fmt.Sprintf("HTTP %d: %s", e.StatusCode, e.Detail)
}

// HandleResponse 统一的API响应处理函数
// HandleResponse parses the template error body and accepts legacy detail.
func HandleResponse(resp *resty.Response, err error, result interface{}) error {
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}

	if resp.StatusCode() >= 300 {
		// Prefer the template message/error fields and fall back to detail.
		var errorData ErrorResponse
		if err := json.Unmarshal(resp.Body(), &errorData); err != nil {
			errorData = ErrorResponse{}
		}
		message := errorData.Message
		if message == "" {
			message = errorData.Detail
		}
		if message == "" {
			message = http.StatusText(resp.StatusCode())
		}
		requestID := resp.Header().Get("X-Request-ID")
		if requestID == "" {
			requestID = errorData.RequestID
		}
		return &HTTPError{
			StatusCode:  resp.StatusCode(),
			Detail:      message,
			Code:        errorData.Code,
			Category:    errorData.Error.Category,
			MachineCode: errorData.Error.Code,
			RequestID:   requestID,
			TraceID:     errorData.TraceID,
			RetryAfter:  resp.Header().Get("Retry-After"),
		}
	}

	if result != nil {
		if err := json.Unmarshal(resp.Body(), result); err != nil {
			return fmt.Errorf("failed to parse response: %w", err)
		}
	}

	return nil
}
